jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPart.java

```diff
@@ -267,20 +346,29 @@ public String toString()
      */
     public static class ChunksPart extends Part
     {
-        private final Content.Source content;
-        private final long length;
+        private final List<Content.Chunk> content;
 
         public ChunksPart(String name, String fileName, HttpFields fields, List<Content.Chunk> content)
         {
             super(name, fileName, fields);
-            this.content = new ChunksContentSource(content);
-            this.length = content.stream().mapToLong(c -> c.getByteBuffer().remaining()).sum();
+            this.content = Objects.requireNonNull(content);
+            content.forEach(Content.Chunk::retain);
+        }
+
+        @Override
+        public Content.Source newContentSource()
+        {
+            List<Content.Chunk> newChunks = content.stream()
+                .map(chunk -> Content.Chunk.from(chunk.getByteBuffer().slice(), chunk.isLast()))
+                .toList();
+            return new ChunksContentSource(newChunks);
```

Comment:
Discussing this with @lorban, I think it's best if we store these `ChunksContentSource` in a `List`, and we fail them all when this part is closed.

The reason is that in this particular case, if a `ChunksContentSource` is read _after_ completing the `Handler` callback, it can read content from buffers that were returned to the pool and may be used for other requests.
By failing them, we guarantee that if they are read after `Handler` completion, an error is returned.

Reply:
What if someone calls `newContentSource()` after the part has been closed, do we need to keep an atomic state and if this has been closed return them a failed instance of `Content.Source` when they call `newContentSource()`?

Why can't we just say if this have been closed it is invalid to use it. Just like if you return a `RetainableByteBuffer` to the pool it is invalid to continue using it.

Reply:
opened #9336